File: src/main/java/mnm/mods/tabbychat/client/gui/settings/GuiSettingsServer.java
package mnm.mods.tabbychat.client.gui.settings;

import mnm.mods.tabbychat.client.TabbyChatClient;
import mnm.mods.tabbychat.client.gui.component.GuiLabel;
import mnm.mods.tabbychat.client.gui.component.config.GuiSettingBoolean;
import mnm.mods.tabbychat.client.gui.component.config.GuiSettingEnum;
import mnm.mods.tabbychat.client.gui.component.config.GuiSettingString;
import mnm.mods.tabbychat.client.gui.component.config.GuiSettingStringList;
import mnm.mods.tabbychat.client.gui.component.config.SettingPanel;
import mnm.mods.tabbychat.client.gui.component.layout.GuiGridLayout;
import mnm.mods.tabbychat.client.settings.GeneralServerSettings;
import mnm.mods.tabbychat.client.settings.ServerSettings;
import mnm.mods.tabbychat.util.ChannelPatterns;
import mnm.mods.tabbychat.util.Color;
import mnm.mods.tabbychat.util.MessagePatterns;
import net.minecraft.client.resources.I18n;
import net.minecraft.util.text.TranslationTextComponent;

import static mnm.mods.tabbychat.util.Translation.*;

public class GuiSettingsServer extends SettingPanel<ServerSettings> {

    GuiSettingsServer() {
        this.setLayout(new GuiGridLayout(10, 20));
        this.setDisplayString(I18n.format(SETTINGS_SERVER));
        this.setSecondaryColor(Color.of(255, 215, 0, 64));
    }

    @Override
    public void initGUI() {
        GeneralServerSettings sett = getSettings().general;

        int pos = 1;
        this.add(new GuiLabel(new TranslationTextComponent(CHANNELS_ENABLED)), new int[]{2, pos});
        GuiSettingBoolean chkChannels = new GuiSettingBoolean(sett.channelsEnabled);
        chkChannels.setCaption(new TranslationTextComponent(CHANNELS_ENABLED_DESC));
        this.add(chkChannels, new int[]{1, pos});

        pos += 1;
        this.add(new GuiLabel(new TranslationTextComponent(PM_ENABLED)), new int[]{2, pos});
        GuiSettingBoolean chkPM = new GuiSettingBoolean(sett.pmEnabled);
        chkPM.setCaption(new TranslationTextComponent(PM_ENABLED_DESC));
        this.add(chkPM, new int[]{1, pos});

        pos += 1;
        add(new GuiLabel(new TranslationTextComponent(USE_DEFAULT)), new int[]{2, pos});
        add(new GuiSettingBoolean(sett.useDefaultTab), new int[]{1, pos});

        pos += 2;
        this.add(new GuiLabel(new TranslationTextComponent(CHANNEL_PATTERN)), new int[]{1, pos});
        GuiSettingEnum<ChannelPatterns> enmChanPat = new GuiSettingEnum<>(sett.channelPattern,
                ChannelPatterns.values());
        enmChanPat.setCaption(new TranslationTextComponent(CHANNEL_PATTERN_DESC));
        this.add(enmChanPat, new int[]{5, pos, 4, 1});

        pos += 2;
        this.add(new GuiLabel(new TranslationTextComponent(MESSAGE_PATTERN)), new int[]{1, pos});
        if (sett.messegePattern.get() == null) {
            sett.messegePattern.set(MessagePatterns.WHISPERS);
        }
        GuiSettingEnum<MessagePatterns> enmMsg = new GuiSettingEnum<>(sett.messegePattern, MessagePatterns.values());
        enmMsg.setCaption(new TranslationTextComponent(MESSAGE_PATTERN_DESC));
        this.add(enmMsg, new int[]{5, pos, 4, 1});

        pos += 2;
        this.add(new GuiLabel(new TranslationTextComponent(IGNORED_CHANNELS)), new int[]{0, pos});
        GuiSettingStringList strIgnored = new GuiSettingStringList(sett.ignoredChannels);
        strIgnored.setCaption(new TranslationTextComponent(IGNORED_CHANNELS_DESC));
        this.add(strIgnored, new int[]{5, pos, 5, 1});

        pos += 2;
        this.add(new GuiLabel(new TranslationTextComponent(DEFAULT_CHANNEL_COMMAND)), new int[]{0, pos});
        GuiSettingString strChannels = new GuiSettingString(sett.channelCommand);
        strChannels.setCaption(new TranslationTextComponent(DEFAULT_CHANNEL_COMMAND_DESC));
        this.add(strChannels, new int[]{5, pos, 5, 1});

        pos += 2;
        this.add(new GuiLabel(new TranslationTextComponent(DEFAULT_CHANNEL)), new int[]{0, pos});
        GuiSettingString strMessages = new GuiSettingString(sett.defaultChannel);
        strMessages.setCaption(new TranslationTextComponent(DEFAULT_CHANNEL_DESC));
        this.add(strMessages, new int[]{5, pos, 5, 1});
    }

    @Override
    public ServerSettings getSettings() {
        return TabbyChatClient.getInstance().getServerSettings();
    }

}

File: src/main/kotlin/mnm/mods/tabbychat/client/gui/settings/GuiSettingsServer.kt
package mnm.mods.tabbychat.client.gui.settings
//
//import mnm.mods.tabbychat.client.TabbyChatClient
//import mnm.mods.tabbychat.client.gui.IScreenHelper
//import mnm.mods.tabbychat.client.gui.Label
//import mnm.mods.tabbychat.client.gui.component.*
//import mnm.mods.tabbychat.client.gui.widget.*
//import mnm.mods.tabbychat.client.settings.ServerSettings
//import mnm.mods.tabbychat.util.*
//import net.minecraft.client.gui.widget.TextFieldWidget
//
//internal class GuiSettingsServer : SettingPanel<ServerSettings> {
//
//    override val settings = TabbyChatClient.serverSettings
//
//    override fun init(screen: IScreenHelper) {
//
//        val left = screen.width / 2 - 80
//        fun top(pos: Int) = pos * 25 + 10
//
//        var pos = 0
////        this.add(Label(Translation.CHANNELS_ENABLED.toComponent()), intArrayOf(2, pos))
//        screen.addWidget(FieldBackedCheckboxButton(settings.general::channelsEnabled,
//                left + 60, top(pos), 80, 20,
//                Translation.CHANNELS_ENABLED.translate())) {
//            //            caption = Translation.CHANNELS_ENABLED_DESC.toComponent()
//        }
//
//        pos += 1
////        this.add(Label(Translation.PM_ENABLED.toComponent()), intArrayOf(2, pos))
//        screen.addWidget(FieldBackedCheckboxButton(settings.general::pmEnabled,
//                left + 60, top(pos), 80, 20,
//                Translation.PM_ENABLED.translate())) {
//            //            caption = Translation.PM_ENABLED_DESC.toComponent()
//        }
//
//        pos += 1
////        this.add(Label(Translation.USE_DEFAULT.toComponent()), intArrayOf(2, pos))
//        screen.addWidget(FieldBackedCheckboxButton(settings.general::useDefaultTab,
//                left + 60, top(pos), 80, 20,
//                Translation.USE_DEFAULT.translate()))
//
//        pos += 1
//        screen.addRender(Label(mc.fontRenderer, Translation.CHANNEL_PATTERN, left, top(pos)))
//        screen.addWidget(SelectWidget(settings.general::channelPattern, EnumValues(ChannelPatterns.values()) { it.display },
//                left + 80, top(pos), 80, 20, "")) {
//            //            caption = Translation.CHANNEL_PATTERN_DESC.toComponent()
//        }
//
//        pos += 1
//        screen.addRender(Label(mc.fontRenderer, Translation.MESSAGE_PATTERN, left, top(pos)))
//        screen.addWidget(SelectWidget(settings.general::messegePattern, EnumValues(MessagePatterns.values()) { it.display },
//                left + 80, top(pos), 80, 20, "")) {
//            //            caption = Translation.MESSAGE_PATTERN_DESC.toComponent()
//        }
//
//        pos += 1
//        screen.addRender(Label(mc.fontRenderer, Translation.IGNORED_CHANNELS, left, top(pos)))
//        screen.addWidget(TextFieldWidget(mc.fontRenderer, left + 80, top(pos), 120, 20, "")) {
//            setResponder { text ->
//                settings.general.ignoredChannels = text.split(",")
//                        .map { it.trim() }
//                        .filter { it.isNotEmpty() }
//            }
//            //            caption = Translation.IGNORED_CHANNELS_DESC.toComponent()
//        }
//
//        // TODO
////        pos += 1
////        this.add(GuiLabel(Translation.DEFAULT_CHANNEL_COMMAND.toComponent()), intArrayOf(0, pos))
////        this.add(GuiSettingString(settings.general::channelCommand), intArrayOf(5, pos, 5, 1)).apply {
////            caption = Translation.DEFAULT_CHANNEL_COMMAND_DESC.toComponent()
////        }
//
//        pos += 2
//        screen.addRender(Label(mc.fontRenderer, Translation.DEFAULT_CHANNEL, left, top(pos)))
//        screen.addWidget(FieldBackedTextFieldWidget(settings.general::defaultChannel,
//                mc.fontRenderer, left + 80, top(pos), 80, 20,
//                Translation.DEFAULT_CHANNEL.translate())) {
//
//            //            caption = Translation.DEFAULT_CHANNEL_DESC.toComponent()
//        }
//    }
//
//}
